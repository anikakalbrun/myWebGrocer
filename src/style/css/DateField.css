.react-datepicker__input-container {
    position: relative;
    display: inline-block;
    width: 100%;
}

.react-datepicker-wrapper {
    display: inline;
}

.valid .react-datepicker__input-container input {
    background-image: url('../images/check.png');
    background-repeat: no-repeat;
    background-position: 98% 50%;
    background-size: 20px;
}

.react-datepicker {
    font-size: 1.3rem;
}

.react-datepicker__header {
    width: 200px;
}

.react-datepicker__month {
    height: 130px;
}

.react-datepicker__current-month, .react-datepicker-time__header {
    font-size: 1.5rem;
}

.react-datepicker__day-name, .react-datepicker__day, .react-datepicker__time-name {
    width: 2.3rem;
}

.react-datepicker__day-names, .react-datepicker__week {
    margin-top: 5px;
}

.react-datepicker__day--keyboard-selected, .react-datepicker__day--selected, .react-datepicker__day--in-selecting-range, .react-datepicker__day--in-range {
    background-color: #43c653;
    color: #fff;
}