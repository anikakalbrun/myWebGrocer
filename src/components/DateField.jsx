import React, {Component} from 'react';
import moment from 'moment';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import {string, bool, shape, func, number} from 'prop-types';

import FieldWrapper from './FieldWrapper';
import '../style/css/DateField.css';
import '../style/css/FormField.css';

export default class DateField extends Component {

    static propTypes = {
        input: shape({
            onChange: func,
            value: string,
        }).isRequired,
        meta: shape({
            touched: bool,
            error: string,
        }).isRequired,
        placeholder: string,
        disabled: bool,
        maxDate: string,
        minDate: string,
        showYearDropdown: bool,
        yearDropdownItemNumber: number
    };

    static defaultProps = {
        showYearDropdown: false,
        yearDropdownItemNumber: 80
    };

    state = {
        selectedDate: this.props.input.value,
    };

    onChange = (date, v) => {
        this.setState({selectedDate: date,});
        const {input, meta: {onChange}} = this.props;
        if (onChange && typeof onChange === 'function') {
            onChange(date, v, input);
        }
    };

    onBlur = (date) => {
        const {input} = this.props;
        input.onBlur(date, input);
    };

    select = () => {
        return this.state.selectedDate ? moment(this.state.selectedDate) : null
    };

    render() {
        const {
            label,
            placeholder,
            disabled,
            maxDate,
            minDate,
            meta: {touched, error},
            showYearDropdown,
            yearDropdownItemNumber
        } = this.props;

        return (
            <FieldWrapper label={label} touched={touched} error={error}>
                <DatePicker selected={this.select()}
                            onChange={this.onChange}
                            onBlur={this.onBlur}
                            className={`form-control ${touched && !error && this.state.selectedDate ? 'valid' : ''}`}
                            maxDate={maxDate && moment(maxDate)}
                            minDate={maxDate && moment(minDate)}
                            placeholderText={placeholder}
                            disabled={disabled}
                            showYearDropdown={showYearDropdown}
                            scrollableYearDropdown
                            yearDropdownItemNumber={yearDropdownItemNumber}
                            dateFormat="DD/MM/YYYY"/>
            </FieldWrapper>
        );
    }
}