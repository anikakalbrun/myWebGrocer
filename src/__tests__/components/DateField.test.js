import React from 'react';
import moment from 'moment';

import DateField from '../../components/DateField';

let wrapper;
const label = "Date";
const placeholder = "Date";
let disabled = false;
const maxDate = '2020-05-21';
const minDate = '1989-09-02';
let meta = {touched: false, error: undefined, onChange: jest.fn()};
const input = {
    onChange: jest.fn(),
    onBlur: jest.fn(),
    value: '2018-06-06'
};


describe('DateField', () => {

    it('renders correctly', () => {
        wrapper = renderer.create(
            <DateField label={label}
                       placeholder={placeholder}
                       disabled={disabled}
                       maxDate={maxDate}
                       minDate={minDate}
                       meta={meta}
                       input={input}
            />).toJSON();
        expect(wrapper).toMatchSnapshot();
    });

    describe('component methods', () => {
        beforeEach(() => {
            meta = {touched: false, error: undefined, onChange: jest.fn()};
            wrapper = shallow(<DateField label={label}
                                         placeholder={placeholder}
                                         disabled={disabled}
                                         maxDate={maxDate}
                                         minDate={minDate}
                                         meta={meta}
                                         input={input}/>
            );
        });

        it('onChange selects proper state and calls props.meta.onChange', () => {
            const instance = wrapper.instance();
            instance.onChange('2019-03-04');
            expect(instance.state.selectedDate).toEqual('2019-03-04');
            expect(instance.props.meta.onChange).toBeCalledWith("2019-03-04", undefined, instance.props.input);
        });

        it('onChange selects proper state and calls props.meta.onChange', () => {
            const instance = wrapper.instance();
            instance.onBlur('2019-03-04');
            expect(instance.props.input.onBlur).toBeCalledWith("2019-03-04", instance.props.input);
        });

        it('select method return proper value', () => {
            const instance = wrapper.instance();
            expect(instance.select()).toEqual(moment(instance.state.selectedDate));
            instance.setState({selectedDate: undefined});
            expect(instance.select()).toEqual(null);
        });
    })

    describe('provides props to FieldWrapper', () => {
        beforeEach(() => {
            meta = {touched: true, error: 'required', onChange: jest.fn()};
            wrapper = shallow(<DateField label={label}
                                         placeholder={placeholder}
                                         disabled={disabled}
                                         maxDate={maxDate}
                                         minDate={minDate}
                                         meta={meta}
                                         input={input}/>
            );
        });

        it('error prop', () => {
            expect(wrapper.find('FieldWrapper').at(0).props().error).toEqual('required');
        })

        it('provides label prop to the FieldWrapper component', () => {
            expect(wrapper.find('FieldWrapper').at(0).props().label).toEqual('Date');
        })

        it('provides touched prop to the FieldWrapper component', () => {
            expect(wrapper.find('FieldWrapper').at(0).props().touched).toEqual(true);
        })
    })
});
