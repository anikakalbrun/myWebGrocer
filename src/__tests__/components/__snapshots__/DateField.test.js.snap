// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateField renders correctly 1`] = `
<div
  className="form-group"
>
  <label
    className="col-xs-12"
  >
    Date
    :
  </label>
  <div
    className="col-xs-12"
  >
    <div>
      <div
        className="react-datepicker-wrapper"
      >
        <div
          className="react-datepicker__input-container"
        >
          <input
            autoComplete={undefined}
            autoFocus={undefined}
            className="form-control "
            disabled={false}
            id={undefined}
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Date"
            readOnly={undefined}
            required={undefined}
            tabIndex={undefined}
            title={undefined}
            type="text"
            value="06/06/2018"
          />
        </div>
      </div>
    </div>
    <div
      className="no-error"
    />
  </div>
</div>
`;
